src/raftstore/store/fsm/peer.rs

```diff
@@ -1835,12 +1833,23 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
                 &mut self.fsm.peer,
             );
         }
+        let target = state.get_target().get_id();
         self.fsm.peer.pending_merge_state = Some(state);
         self.notify_prepare_merge();
 
-        if merged {
-            // CommitMerge will try to catch up log for source region. If PrepareMerge is executed
-            // in the progress of catching up, there is no need to schedule merge again.
+        if let Some(logs_up_to_date) = self.fsm.peer.catch_up_logs.take() {
+            logs_up_to_date.store(region.get_id(), Ordering::SeqCst);
+            let mailbox = self.ctx.apply_router.mailbox(target).unwrap();
+            // Send CatchUpLogs back to destroy source apply delegate,
+            // then it will send `LogsUpToDate` to target apply delegate.
+            self.ctx.apply_router.schedule_task(
```

Comment:
What if there are some cascaded merges? Can this region be destroyed?

Reply:
see https://github.com/tikv/tikv/pull/4595/files/b15fcb95aa35aecdddf632aecf108ed0336875b3#diff-641634cfd274be1752afcbf3c83d9200L2121, if there are some cascaded merges, it will return directly. And `on_ready_prepare_merge` is not called until all the cascaded merges are finished.